### ISSUE TITLE
[DeepEP] How is _handle_cache handled during inference?

### ISSUE THREAD

**Issue by goldhuang** [pytorch/torchtitan#2273]
When I search `_handle_cache` in the code base, I only found the lines below. Seems the cache is not properly released if I do inference in between training steps, like inference to get validation loss etc., because `_handle_cache.pop(ctx.cache_id_int, None)` is only in dispatch_bwd.

```
_handle_cache: dict = {}
_cache_counter: int = 0
 
 
 
_handle_cache[cache_id.item()] = handle
 
handle = _handle_cache.get(cache_id.item())
 
handle = _handle_cache.get(ctx.cache_id_int)
 
_handle_cache.pop(ctx.cache_id_int, None)
 
ctx.saved_handle = _handle_cache.get(ctx.cache_id_int)

```
@elfiegg

**Comment by tianyu-l** [comment]
cc @elfiegg @shuhuayu

**Comment by elfiegg** [comment]
This is by design - cache_id is to save the `handle` returned by forward dispatch API call, and this `handle` will be used for the following combine and both processes' backward. So the life cycle would be:
```
dispatch generates handle and cache_id -> saves id to ctx.id -> combine retrieves handle from cache -> saves handle to ctx.handle.
dispatch backward retrieves handle via cache.get(ctx.id) -> release cache
combine backward retrieves handle via ctx.handle
```
Since we need to access handle across ops (dispatch and combine), as well as in both ops' backwards - we can't clean up the cache immediately after dispatch, but it's possible to save handle in dispatch_setup_context and clean it up after combine_setup_context. So the flow could potentially be:
```
dispatch generates handle and cache_id -> saves id to ctx.id, also saves handle to ctx.handle -> combine retrieves handle from cache -> saves handle to ctx.handle, and release the cache.
dispatch backward retrieves handle via ctx.handle
combine backward retrieves handle via ctx.handle
```
Do you run into any issue with the current implementation?

References:
1. DeepEP implementation: https://github.com/deepseek-ai/DeepEP?tab=readme-ov-file#example-use-in-model-training-or-inference-prefilling
2. HybridEP deisgn doc, "Handle Cache Strategy": https://docs.google.com/document/d/1i8zlu-3S2psDztKc9hHuyRAu2tsskDRBmMl3egDo7Vo/edit?tab=t.0#heading=h.t9wg1x5hf31w

**Comment by goldhuang** [comment]
@elfiegg I see memory leak when we only run forward, e.g. computing validation loss. Seems the current implementation enforces fwd+bwd+sac. Cache in _handle_cache cannot be released when we only run fwd.

I'm planning to reset _handle_cache and _cache_counter after each forward when computing validation loss. Do you see any issue if I do like that?

**Comment by elfiegg** [comment]
You should be able to save handle in `_dispatch_setup_context` and free the cache in `_combine_setup_context` without issues like this draft: https://github.com/pytorch/torchtitan/pull/2274/files

I've quickly run it on 16B model and it worked.

**Comment by shuhuayu** [comment]
@goldhuang Thanks for the issue. Yes, there was a memory leak issue with the current implementation. You can refer to this branch for an ad hoc fix: https://github.com/shuhuayu/torchtitan/tree/deepep, which has been tested on a 512 gpu run. We will update an clean fix soon.
